pkg/registry/populator_test.go

```diff
@@ -1534,3 +1534,759 @@ func TestSemverPackageManifest(t *testing.T) {
 		})
 	}
 }
+
+func TestSubstitutesFor(t *testing.T) {
+	type args struct {
+		bundles  []string
+		rebuilds []string
+	}
+	type replaces map[string]string
+	type expected struct {
+		bundles        []string
+		channels       []string
+		substitutions  map[string]string
+		whatReplaces   map[string]replaces
+		channelHeads   map[string]string
+		defaultChannel string
+	}
+	tests := []struct {
+		description string
+		args        args
+		expected    expected
+	}{
+		{
+			description: "SubstitutesFor/FirstSubstitutionReplaces",
+			args: args{
+				bundles: []string{
+					"prometheus.0.22.2",
+					"prometheus.0.14.0",
+					"prometheus.0.15.0",
+				},
+				rebuilds: []string{
+					"prometheus.0.15.0.substitutesfor",
+				},
+			},
+			expected: expected{
+				bundles: []string{
+					"quay.io/test/prometheus.0.22.2",
+					"quay.io/test/prometheus.0.14.0",
+					"quay.io/test/prometheus.0.15.0.substitutesfor",
+					"quay.io/test/prometheus.0.15.0.substitutesfor",
+					"quay.io/test/prometheus.0.15.0",
+					"quay.io/test/prometheus.0.15.0",
+					"quay.io/test/prometheus.0.14.0",
+				},
+				channels: []string{
+					"stable",
+					"preview",
+					"preview",
+					"stable",
+					"stable",
+					"preview",
+					"preview",
+				},
+				substitutions: map[string]string{
+					"prometheusoperator.0.22.2":                      "",
+					"prometheusoperator.0.14.0":                      "",
+					"prometheusoperator.0.15.0":                      "",
+					"prometheusoperator.0.15.0+1-freshmaker-rebuild": "prometheusoperator.0.15.0",
+				},
```

Comment:
Do you actually need any of this when you have already have `whatReplaces`, `channelHeads`, and `defaultChannel` representing the effective graph already?

Reply:
Makes sense! Let's see what other reviewers think.